lib/segment/src/index/field_index/numeric_index/mod.rs

```diff
@@ -196,26 +197,50 @@ impl<T: Encodable + Numericable + Default> NumericIndex<T> {
         }
     }
 
-    fn range_cardinality(&self, range: &Range<FloatPayloadType>) -> CardinalityEstimation {
+    fn range_cardinality(&self, range: &RangeInterface) -> CardinalityEstimation {
         let max_values_per_point = self.max_values_per_point();
         if max_values_per_point == 0 {
             return CardinalityEstimation::exact(0);
         }
 
-        let lbound = if let Some(lte) = range.lte {
-            Included(T::from_f64(lte))
-        } else if let Some(lt) = range.lt {
-            Excluded(T::from_f64(lt))
-        } else {
-            Unbounded
-        };
+        let (lbound, gbound) = match range {
+            RangeInterface::Float(float_range) => {
+                let lbound = if let Some(lte) = float_range.lte {
+                    Included(T::from_f64(lte))
+                } else if let Some(lt) = float_range.lt {
+                    Excluded(T::from_f64(lt))
+                } else {
+                    Unbounded
+                };
 
-        let gbound = if let Some(gte) = range.gte {
-            Included(T::from_f64(gte))
-        } else if let Some(gt) = range.gt {
-            Excluded(T::from_f64(gt))
-        } else {
-            Unbounded
+                let gbound = if let Some(gte) = float_range.gte {
+                    Included(T::from_f64(gte))
+                } else if let Some(gt) = float_range.gt {
+                    Excluded(T::from_f64(gt))
+                } else {
+                    Unbounded
+                };
+                (lbound, gbound)
+            }
+            // TODO: Maybe refactor `range_cardinality` out of `PayloadFieldIndex` to allow polymorphism, as in `ValueIndexer<T>`
+            RangeInterface::DateTime(datetime_range) => {
+                let lbound = if let Some(lte) = datetime_range.lte {
+                    Included(T::from_i64(lte.timestamp_micros()))
+                } else if let Some(lt) = datetime_range.lt {
+                    Excluded(T::from_i64(lt.timestamp_micros()))
+                } else {
+                    Unbounded
+                };
+
+                let gbound = if let Some(gte) = datetime_range.gte {
+                    Included(T::from_i64(gte.timestamp_micros()))
+                } else if let Some(gt) = datetime_range.gt {
+                    Excluded(T::from_i64(gt.timestamp_micros()))
+                } else {
+                    Unbounded
+                };
+                (lbound, gbound)
+            }
```

Comment:
You could reduce code duplication by converting to `Range<T>` first.
Also, an idea: add `to_bounds` method to the `Range<T>` structure.

```rust
impl<T: Copy> Range<T> {
    fn to_bounds(&self) -> (Bound<T>, Bound<T>) {...}
}
```

So this code could be reduced to this:
```rust
let (lbound, gbound) = match range {
    RangeInterface::Float(range) => range.map(T::from_f64),
    RangeInterface::DateTime(range) => range.map(|t| T::from_i64(i.timestamp_micros())),
}.to_bounds();
```

Reply:
I actually implemented a similar thing for #3493, I'm going to reuse it here 👍